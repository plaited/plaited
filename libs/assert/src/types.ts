export interface AssertionErrorInterface extends Error {
  name: string
}

export interface AssertionErrorConstructor {
  new (message: string): AssertionErrorInterface
}
