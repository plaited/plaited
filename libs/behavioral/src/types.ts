export interface StateSnapshot {
  (props: { bids: PendingBid[]; selectedEvent: CandidateBid }): {
    thread: string
    request?: RequestIdiom[]
    waitFor?: ParameterIdiom[]
    block?: ParameterIdiom[]
    priority: number
  }[]
}

export type Detail = unknown | (() => unknown) | Event

export type SnapshotMessage = ReturnType<StateSnapshot>

export type SelectedMessage = {
  type: string
  detail?: Detail
}

export type Trigger = <T extends Detail = Detail>(args: TriggerArgs<T>) => void

export type TriggerArgs<T extends Detail = Detail> = {
  type: string
  detail?: T
}
// Rule types
type Callback<T extends Detail = Detail> = (args: { type: string; detail: T }) => boolean

export type ParameterIdiom<T extends Detail = Detail> =
  | {
      type: string
      cb?: Callback<T>
    }
  | {
      type?: string
      cb: Callback<T>
    }

export type RequestIdiom<T extends Detail = Detail> = {
  type: string
  detail?: T
}

export type RuleSet<T extends Detail = Detail> = {
  waitFor?: ParameterIdiom<T> | ParameterIdiom<T>[]
  request?: RequestIdiom<T> | RequestIdiom<T>[]
  block?: ParameterIdiom<T> | ParameterIdiom<T>[]
}

export type RulesFunc<T extends Detail = Detail> = () => IterableIterator<RuleSet<T>>

export type RunningBid = {
  trigger?: true
  thread: string
  priority: number
  generator: IterableIterator<RuleSet>
}
export type PendingBid = RuleSet & RunningBid

export type CandidateBid = {
  priority: number
  type: string
  detail?: Detail
  cb?: Callback
}

export type Strategy = (filteredEvents: CandidateBid[] | never[]) => CandidateBid | undefined

// Feedback Types
type Actions<T extends Record<string, (detail: Detail) => void>> = {
  [K in keyof T]: T[K] extends (detail: infer D) => void ? (detail: D extends Detail ? D : Detail) => void : never
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Feedback = <T extends Record<string, (detail: any) => void>>(actions: Actions<T>) => void

export interface DevCallback {
  (args: ReturnType<StateSnapshot>): void
}

export type Log = ReturnType<StateSnapshot>
