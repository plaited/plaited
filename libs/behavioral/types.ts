import { streamEvents } from './constants.ts'

export interface StateSnapshot {
  (
    props: {
      bids: PendingBid[]
      selectedEvent: CandidateBid
    },
  ): {
    thread: string
    request?: RequestIdiom[]
    waitFor?: ParameterIdiom[]
    block?: ParameterIdiom[]
    priority: number
  }[]
}
export type StreamMessage = {
  type: typeof streamEvents.select
  data: {
    event: string
    detail?: Record<string, unknown>
  }
} | {
  type: typeof streamEvents.snapshot
  data: ReturnType<StateSnapshot>
}

export type Listener = (msg: StreamMessage) => StreamMessage | void

export interface Stream {
  (value: StreamMessage): void
  subscribe: (listener: Listener) => Stream
}

export type Trigger = <
  T extends Record<string, unknown> = Record<string, unknown>,
>(args: {
  event: string
  detail?: T
}) => void

export type TriggerArgs = Parameters<Trigger>[0]

// Rule types
type Callback<
  T extends (Record<string, unknown> | Event) = Record<string, unknown>,
> = (
  args: { event: string; detail: T },
) => boolean

export type ParameterIdiom<
  T extends (Record<string, unknown> | Event) = Record<string, unknown>,
> = {
  event: string
  cb?: Callback<T>
} | {
  event?: string
  cb: Callback<T>
}

export type RequestIdiom<
  T extends (Record<string, unknown> | Event) = Record<string, unknown>,
> = {
  event: string
  detail?: T
  // cb?: Callback<T>
}

export type RuleSet<
  T extends (Record<string, unknown> | Event) = Record<string, unknown>,
> = {
  waitFor?: ParameterIdiom<T> | ParameterIdiom<T>[]
  request?: RequestIdiom<T> | RequestIdiom<T>[]
  block?: ParameterIdiom<T> | ParameterIdiom<T>[]
}

export type RulesFunc<
  T extends (Record<string, unknown> | Event) = Record<string, unknown>,
> = () => IterableIterator<
  RuleSet<T>
>

export type RunningBid = {
  trigger?: true
  thread: string
  priority: number
  generator: IterableIterator<RuleSet>
}
export type PendingBid = RuleSet & RunningBid

export type CandidateBid = {
  priority: number
  event: string
  detail?: Record<string, unknown>
  cb?: Callback
}

export type Strategy = (
  filteredEvents: CandidateBid[] | never[],
) => CandidateBid | undefined

// Feedback Types
type Actions<
  T extends Record<string, (detail: Record<string, unknown>) => void>,
> = {
  [K in keyof T]: T[K] extends (detail: infer D) => void ? (
      detail: D extends Record<string, unknown> ? D : Record<string, unknown>,
    ) => void
    : never
}

// deno-lint-ignore no-explicit-any
export type Feedback = <T extends Record<string, (detail: any) => void>>(
  actions: Actions<T>,
) => void

export interface DevCallback {
  (args: ReturnType<StateSnapshot>): void
}
