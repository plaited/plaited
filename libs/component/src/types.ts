import {
  bProgram,
  DevCallback,
  Strategy,
  Trigger,
  TriggerArgs,
} from '@plaited/behavioral'
import { SugaredElement } from './sugar.js'
import { Template } from '@plaited/jsx'

export type Plait = (props: PlaitProps) => void | Promise<void>

export type PlaitedElementOptions = {
  /** define wether island's custom element is open or closed. @defaultValue 'open'*/
  mode?: 'open' | 'closed';
  /** configure whether to delegate focus or not @defaultValue 'true' */
  delegatesFocus?: boolean;
  /** logger function to receive messages from behavioral program react streams */
  dev?: DevCallback;
  /** event selection strategy callback from behavioral library */
  strategy?: Strategy;
  /** messenger connect callback */
  connect?: (recipient: string, trigger: Trigger) => () => void;
  /** set to true if we wish to use id when connecting to messenger to receive messages from other islands */
  id?: boolean;
  /** the element tag you want to use */
  tag: `${string}-${string}`;
};
export interface PlaitedElement extends HTMLElement {
  internals_: ElementInternals
  plait?(props: PlaitProps): void | Promise<void>;
  connectedCallback?(): void;
  attributeChangedCallback?(
    name: string,
    oldValue: string | null,
    newValue: string | null,
  ): void;
  disconnectedCallback?(): void;
  adoptedCallback?(): void;
  formAssociatedCallback?(form: HTMLFormElement): void;
  formDisabledCallback?(disabled: boolean): void;
  formResetCallback?(): void;
  formStateRestoreCallback?(
    state: unknown,
    reason: 'autocomplete' | 'restore',
  ): void;
}
export type SelectorMod = '=' | '~=' | '|=' | '^=' | '$=' | '*='
export type PlaitProps = {
  /** query for elements with the data-target attribute in the Island's shadowDom and slots */
  $<T extends HTMLElement | SVGElement = HTMLElement | SVGElement>(
    target: string,
    opts?: {
      all?: false;
      mod?: SelectorMod;
    }
  ): SugaredElement<T> | undefined;
  $<T extends HTMLElement | SVGElement = HTMLElement | SVGElement>(
    target: string,
    /** This options enables querySelectorAll and modified the attribute selector for data-target{@default {all: false, mod: "=" } } {@link https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors#syntax}*/
    opts?: {
      all: true;
      mod?: SelectorMod;
    },
  ): SugaredElement<T>[];
  /** The DOM node context allowing easy light & shadow dom access
   * @example
   * // returns the div element inside
   * // the shadowRoot of the element instance
   * const shadowEl = host.shadowRoot.querySelector('div')
   */
  host: PlaitedElement;
} & ReturnType<typeof bProgram>;

export interface PlaitedElementConstructor {
  template?: Template;
  observedTriggers?: Record<string, string>
  new (): PlaitedElement;
}

export type SendMessage = (recipient: string, detail: TriggerArgs) => void;
export type BroadcastMessage = (recipient: TriggerArgs) => void;


export type Position = 'beforebegin' | 'afterbegin' | 'beforeend' | 'afterend';

export type PlaitedComponent = {
  (): void;
  tag: string;
}

export type CreateComponent =  ({
  mode,
  delegatesFocus,
  tag,
  ...bProgramOptions
}: PlaitedElementOptions, mixin?: (base: PlaitedElementConstructor) => PlaitedElementConstructor) => PlaitedComponent
