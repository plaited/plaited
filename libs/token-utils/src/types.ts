/**
 * Alias value type points to another token value
 */
export type AliasValue = `{${string}}`

/**
 * Default value type relies on the defaultFormat formatter for ts tokens and css tokens
 */
export type DefaultValue = string | number | AliasValue | (string | number | AliasValue)[] 

/**
 * Number or percentage values relies on the defaultFormat formatter for ts tokens and css tokens
 */
export type AmountValue = number | `${number}%` | AliasValue

/**
 * Number or percentage values relies on the defaultFormat formatter for ts tokens and css tokens
 */
export type AngleValue = `${number}deg` | `${number}grad` | `${number}rad` | `${number}turn` | AliasValue

/**
 * Color value type relies on the color formatter for css tokens and defaultFormat formatter for ts tokens
 */
export type ColorValue =
  | {
      l?: AmountValue
      c?: AmountValue
      h?: number | AngleValue
      a?: AmountValue
    }
  | `#${string}`
  |AliasValue


/**
 * Size value type relies on the size formatter for css tokens
 * and the defaultFormat formatter for ts tokens
 */
export type SizeValue = AmountValue | AmountValue[]

/**
 * Gradient value type relies on the gradient formatter for css tokens
 * and the defaultFormat formatter for ts tokens
 */
export type GradientValue =
  | {
      gradientFunction:
        | 'linear-gradient'
        | 'radial-gradient'
        | 'conic-gradient'
        | 'repeating-linear-gradient'
        | 'repeating-radial-gradient'
        | 'repeating-conic-gradient'
      angleShapePosition?: string
      colorStops: {
        color?: ColorValue
        position?: string
      }[]
    }
  | AliasValue

export type CompositeValue =
  | { [key: string]: AliasValue}
  | AliasValue

export type ContextValue<V> = {
  [key: string]: V
}

export type DesignValue =
  | DefaultValue
  | ColorValue
  | SizeValue
  | GradientValue
  | CompositeValue

export type ContextTypes = 'media-query' | 'color-scheme'

export type StaticToken<V extends DesignValue, T = undefined> = {
  $description: string
  $extensions?: {
    ['plaited']?: {
      context?: never
      commaSeparated?: boolean
    }
    [key: string]: unknown
  }
  $value: V
} & (T extends undefined ? { $type?: never } : { $type: T })

export type ContextualToken<V extends DesignValue, T = undefined> = {
  $description: string
  $extensions: {
    ['plaited']: {
      context: ContextTypes
      commaSeparated?: boolean
    }
    [key: string]: unknown
  }
  $value: ContextValue<V>
} & (T extends undefined ? { $type?: never } : { $type: T })

export type BaseToken<V extends DesignValue, T = undefined> = StaticToken<V, T> | ContextualToken<V, T>

export type DefaultToken = BaseToken<DefaultValue>

export type AmountToken = BaseToken<AmountValue, 'amount'> 

export type AngleToken = BaseToken<AngleValue, 'angle'>

export type ColorToken = BaseToken<ColorValue, 'color'>

export type SizeToken = BaseToken<SizeValue, 'size'>

export type GradientToken = BaseToken<GradientValue, 'gradient'>

export type CompositeToken = {
  $description: string
  $extensions?: {
    [key: string]: unknown
  }
  $type: 'composite'
  $value: CompositeValue
}

export type DesignToken =
  | DefaultToken
  | AmountToken
  | AngleToken
  | ColorToken
  | SizeToken
  | GradientToken
  | CompositeToken

// general tokens object type definition
export type DesignTokenGroup = {
  [key: string]: DesignTokenGroup | DesignToken
}

export type MediaQueries = {
  [key: string]: string;
};

export type ColorSchemes = {
  light?: "light";
  dark?: "dark";
};

export type Contexts = {
  mediaQueries: MediaQueries;
  colorSchemes: ColorSchemes;
};

export type FormatList = (args: {
  tokens: DesignTokenGroup;
  allTokens: DesignTokenGroup;
  tokenPath?: string[];
  getFormatters: GetFormatters;
  contexts: Contexts;
  $type?: string,
}) => string;

export type Formatter<T extends DesignToken = DesignToken> = (
  token: T,
  details: {
    tokenPath: string[];
    allTokens: DesignTokenGroup;
    contexts: Contexts;
  }
) => string;

export type GetFormatters = <
  T extends DesignTokenGroup = DesignTokenGroup,
  F extends DesignToken = DesignToken,
>(
  token: F,
  details: {
    tokenPath: string[];
    allTokens: T;
    contexts: Contexts;
  }
) => string;

export type TransformerParams = {
  /** an object of the type {@link DesignTokenGroup} */
  tokens: DesignTokenGroup;
  contexts: Contexts;
};

export type TransformTokensParams = {
  /** an object of the type {@link DesignTokenGroup} */
  tokens: DesignTokenGroup;
  /** named media queries */
  contexts?: Partial<Contexts>;
};