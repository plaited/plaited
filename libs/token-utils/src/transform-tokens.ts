import { transformToCSS } from "./transformers/transform-to-css.js";
import { transformToTS } from "./transformers/transform-to-ts.js";
import { TransformTokensParams } from "./types.js";

export const transformTokens = ({
  tokens,
  baseFontSize = 20,
  contexts: { mediaQueries = {}, colorSchemes = {} } = {},
}: TransformTokensParams): {
  css: string;
  ts: string;
} => ({
  css: transformToCSS({
    tokens,
    baseFontSize,
    contexts: {
      colorSchemes,
      mediaQueries,
    },
  }),
  ts: transformToTS({
    tokens,
    baseFontSize,
    contexts: {
      colorSchemes,
      mediaQueries,
    },
  }),
});