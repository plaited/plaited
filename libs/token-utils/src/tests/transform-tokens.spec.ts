import { test, expect, jest } from 'bun:test'
import beautify from 'beautify'

import { transformTokens } from "../transform-tokens.js";
import { DesignTokenGroup } from "../types.js";

test("empty token group", () => {
  const tokens = {
    colors: {},
  };
  const { css, ts } = transformTokens({
    tokens,
  });
  expect(beautify(css, { format: 'css' })).toBe('');
  expect(ts).toBe('');
});

test("token group", () => {
  const tokens: DesignTokenGroup = {
    lineHeight: {
      xxs: { $value: 0.25, $type: "lineHeight", $description: "mock description" },
      xs: { $value: 0.5, $type: "lineHeight", $description: "mock description" },
      sm: { $value: 0.75, $type: "lineHeight", $description: "mock description" },
      md: { $value: 1, $type: "lineHeight", $description: "mock description" },
      ml: { $value: 1.25, $type: "lineHeight", $description: "mock description" },
      lg: { $value: 1.5, $type: "lineHeight", $description: "mock description" },
      xl: { $value: 2, $type: "lineHeight", $description: "mock description" },
      xxl: { $value: 3, $type: "lineHeight", $description: "mock description" },
      xxxl: { $value: 3.5, $type: "lineHeight", $description: "mock description" },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("nested token group", () => {
  const tokens: DesignTokenGroup = {
    colors: {
      gray: {
        10: { $value: {
          l: "97.91%",
          c: 0,
          h: 0
        }, $type: "color", $description: "mock description" },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("context: [media-query]", () => {
  const tokens: DesignTokenGroup = {
    size: {
      dynamic: {
        100: {
          $value: {
            desktop: 3,
            tv: 4,
            mobile: 2,
          },
          $type: "dimension",
          $description: "mock description",
          $extensions: { "plaited-context": "media-query" },
        },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
    contexts: {
      mediaQueries: {
        desktop: "screen and (min-width: 1024px)",
        tv: "screen and (min-width: 1920px)",
        mobile: "screen and (max-width: 767px)",
      },
    },
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("context: [color-scheme]", () => {
  const tokens: DesignTokenGroup = {
    colors: {
      background: {
        primary: {
          $value: {
            light: {
              l: "100%",
              c: 0,
              h: 0,
              a: 0.5,
            },
            dark: {
              l: "0%",
              c: 0,
              h: 0,
              a: 0.5,
            },
          },
          $type: "color",
          $extensions: { "plaited-context": "color-scheme" },
          $description: "mock description",
        },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
    contexts: {
      colorSchemes: {
        light: "light",
        dark: "dark",
      },
    },
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("context: [media-query, color-scheme]", () => {
  const tokens: DesignTokenGroup = {
    size: {
      dynamic: {
        100: {
          $value: {
            tv: 4,
            mobile: 2,
            desktop:3,
          },
          $type: "dimension",
          $description: "mock description",
          $extensions: { "plaited-context": "media-query" },
        },
      },
    },
    colors: {
      background: {
        primary: {
          $value: {
            light: {
              l: "100%",
              c: 0,
              h: 0,
              a: 0.5,
            },
            dark: {
              l: "0%",
              c: 0,
              h: 0,
              a: 0.5,
            },
          },
          $type: "color",
          $extensions: { "plaited-context": "color-scheme" },
          $description: "mock description",
        },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
    contexts: {
      mediaQueries: {
        desktop: "screen and (min-width: 1024px)",
        mobile: "screen and (max-width: 767px)",
        tv: "screen and (min-width: 1920px)",
      },
      colorSchemes: {
        light: "light",
        dark: "dark",
      },
    },
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("alias", () => {
  const tokens: DesignTokenGroup = {
    colors: {
      gray: {
        100: { $value: {
          l: "98.21%",
          c: 0,
          h: 0
        }, $type: "color", $description: "mock description" },
      },
    },
    surface: {
      secondary: {
        100: {
          $value: "{colors.gray.100}",
          $type: "color",
          $description: "mock description"
        },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("alias + context", () => {
  const tokens: DesignTokenGroup = {
    colors: {
      white: {
        100: { $value: {
          l: "97.91%",
          c: 0,
          h: 0
        }, $type: "color", $description: "mock description" },
      },
      charcoal: {
        100: { $value: {
          l: "18.31%",
          c: 0.004,
          h: 285.99
        }, $type: "color", $description: "mock description" },
      },
    },
    surface: {
      secondary: {
        100: {
          $value: {
            light: "{colors.white.100}",
            dark: "{colors.charcoal.100}",
          },
          $type: "color",
          $description: "mock description",
          $extensions: { "plaited-context": "color-scheme" },
        },
      },
    },
  };
  const { css, ts } = transformTokens({
    tokens,
    contexts: {
      colorSchemes: {
        light: "light",
        dark: "dark",
      },
    },
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("exercise token types", () => {
  const tokens: DesignTokenGroup = {
    fontWeight: { $value: 700, $type: "fontWeight", $description: "mock description" },
    letterSpacing: {
      $value: "normal",
      $type: "primitive",
      $description: "mock description"
    },
    border: {
      $value: {
        width: 1,
        style: "solid",
        color: {
          l: "0%",
          c: 0,
          h: 0,
        },
      },
      $type: "border",
      $description: "mock description"
    },
    dropShadow: {
      $value: {
        offsetX: 0,
        offsetY: 4,
        blur: 4,
        color: {
          l: "0%",
          c: 0,
          h: 0,
          a: 0.5,
        },
      },
      $type: "dropShadow",
      $description: "mock description"
    },
    gradient: {
      $value: {
        gradientFunction: "linear-gradient",
        angleShapePosition: "45deg",
        colorStops: [
          { color: {
            l: "0%",
            c: 0,
            h: 0,
            a: 0.5,
          }, position: "20%" },
          { color: {
            l: "100%",
            c: 0,
            h: 0,
            a: 0.5,
          }, position: "80%" },
        ],
      },
      $type: "gradient",
      $description: "mock description"
    },
    fontFamily: {
      $value: ["Roboto", "sans-serif"],
      $type: "fontFamily",
      $description: "mock description"
    },
    transition: {
      $value: {
        duration: "0.3s",
        delay: "0s",
        timingFunction: "ease-in-out",
      },
      $type: "transition",
      $description: "mock description"
    },
    transitionCubic: {
      $value: {
        duration: "0.3s",
        delay: "0s",
        timingFunction: {
          function: "cubic-bezier",
          values: [0.25, 0.1, 0.25, 1],
        },
      },
      $type: "transition",
      $description: "mock description"
    },
    gap: {
      $value: 10,
      $type: "gap",
      $description: "mock description"
    },
    gapPercent: {
      $value: "10%",
      $type: "gap",
      $description: "mock description"
    },
    gridTemplate: {
      $value: `"a a a"\n"b c c"\n"b c c"`,
      $type: "gridTemplate",
      $description: "mock description"
    },
    gridTemplateColumns: {
      $value: [60, 60],
      $type: "gridTemplate",
      $description: "mock description"
    },
    fontSize: {
      $value: 14,
      $type: "dimension",
      $description: "mock description",
    },
    lineHeight: {
      $value: 1,
      $type: "primitive",
      $description: "mock description"
    },
    typography: {
      $value: {
        fontFamily: "{fontFamily}",
        fontSize: "{fontSize}",
        fontWeight: "{fontWeight}",
        letterSpacing: "{letterSpacing}",
        lineHeight: "{lineHeight}",
      },
      $type: "typography",
      $description: "mock description"
    },
    flex: {
      $value: {
        display: "flex",
        columnGap: "{gap}",
      },
      $type: "flex",
      $description: "mock description"
    },
    grid: {
      $value: {
        display: "inline-grid",
        columnGap: "{gap}",
        rowGap: "{gapPercent}",
      },
      $type: "grid",
      $description: "mock description"
    },
    primitiveString: {
      $value: "primitive string",
      $type: "primitive",
      $description: "mock description"
    },
    primitiveNumber: {
      $value: 50,
      $type: "primitive",
      $description: "mock description"
    },
  };
  const { css, ts } = transformTokens({
    tokens,
    contexts: {
      colorSchemes: {
        light: "light",
        dark: "dark",
      },
    },
  });
  expect(beautify(css, { format: 'css' })).toMatchSnapshot();
  expect(ts).toMatchSnapshot();
});

test("invalid alias", () => {
  const tokens: DesignTokenGroup = {
    colors: {
      white: {
        100: {
          $value: {
            l: "100%",
            c: 0,
            h: 0,
            a: 0.5,
          },
          $type: "color",
          $description: "mock description"
        },
      },
    },
    surface: {
      secondary: {
        100: {
          $value: "{colors.white.200}",
          $type: "color",
          $description: "mock description"
        },
      },
    },
  };
  console.error = jest.fn();
  transformTokens({
    tokens,
  });
  //@ts-expect-error: it exist
  console.error.mock.calls.forEach((call) => {
    expect(call[0]).toBe(`Invalid token alias: {colors.white.200}`);
  });
});

test("invalid context", () => {
  const tokens: DesignTokenGroup = {
    size: {
      dynamic: {
        "1": {
          $value: {
            desktop: 2,
          },
          $type: "dimension",
          $description: "mock description",
          //@ts-expect-error: it exist
          $extensions: { "plaited-context": "media" },
        },
      },
    },
  };
  transformTokens({
    tokens,
  });
  console.error = jest.fn();
  transformTokens({
    tokens,
  });
  //@ts-expect-error: it exist
  console.error.mock.calls.forEach((call) => {
    expect(call[0]).toBe(`Context type [media] is an invalid context type`);
  });
});

test("invalid context key", () => {
  const tokens: DesignTokenGroup = {
    size: {
      dynamic: {
        "1": {
          $value: {
            tv: 4,
          },
          $type: "dimension",
          $description: "mock description",
          $extensions: { "plaited-context": "media-query" },
        },
      },
    },
  };
  console.error = jest.fn();
  transformTokens({
    tokens,
    contexts: {
      mediaQueries: {
        desktop: "screen and (min-width: 1024px)",
        mobile: "screen and (max-width: 767px)",
      },
    },
  });
  //@ts-expect-error: it exist
  console.error.mock.calls.forEach((call) => {
    expect(call[0]).toBe(`[tv] not found in mediaQueries`);
  });
});