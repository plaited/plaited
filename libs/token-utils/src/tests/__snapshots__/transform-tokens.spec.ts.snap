// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`token group 1`] = `
":host {
    --line-height-xxs: 0.0125rem;
    --line-height-xs: 0.025rem;
    --line-height-sm: 0.0375rem;
    --line-height-md: 0.05rem;
    --line-height-ml: 0.0625rem;
    --line-height-lg: 0.075rem;
    --line-height-xl: 0.1rem;
    --line-height-xxl: 0.15rem;
    --line-height-xxxl: 0.175rem;
}"
`;

exports[`token group 2`] = `
"export const lineHeightXxs = 'var(--line-height-xxs)'
export const lineHeightXs = 'var(--line-height-xs)'
export const lineHeightSm = 'var(--line-height-sm)'
export const lineHeightMd = 'var(--line-height-md)'
export const lineHeightMl = 'var(--line-height-ml)'
export const lineHeightLg = 'var(--line-height-lg)'
export const lineHeightXl = 'var(--line-height-xl)'
export const lineHeightXxl = 'var(--line-height-xxl)'
export const lineHeightXxxl = 'var(--line-height-xxxl)'
"
`;

exports[`nested token group 1`] = `
":host {
    --colors-gray-10: oklch(97.91% 0 0 / none);
}"
`;

exports[`nested token group 2`] = `
"export const colorsGray10 = 'var(--colors-gray-10)'
"
`;

exports[`context: [media-query] 1`] = `
":host {
    --size-dynamic-100: 0.15rem;
}

@media screen and (min-width: 1024px) {
    :host {
        --size-dynamic-100: 0.15rem;
    }
}

:host([data-media-query="desktop"]) {
    --size-dynamic-100: 0.15rem;
}

@media screen and (min-width: 1920px) {
    :host {
        --size-dynamic-100: 0.2rem;
    }
}

:host([data-media-query="tv"]) {
    --size-dynamic-100: 0.2rem;
}

@media screen and (max-width: 767px) {
    :host {
        --size-dynamic-100: 0.1rem;
    }
}

:host([data-media-query="mobile"]) {
    --size-dynamic-100: 0.1rem;
}"
`;

exports[`context: [media-query] 2`] = `
"export const sizeDynamic100 = 'var(--size-dynamic-100)'
"
`;

exports[`context: [color-scheme] 1`] = `
":host {
    --colors-background-primary: oklch(100% 0 0 / 0.5);
}

@media (prefers-color-scheme:light) {
    :host {
        --colors-background-primary: oklch(100% 0 0 / 0.5);
    }
}

:host([data-color-scheme="light"]) {
    --colors-background-primary: oklch(100% 0 0 / 0.5);
}

@media (prefers-color-scheme:dark) {
    :host {
        --colors-background-primary: oklch(0% 0 0 / 0.5);
    }
}

:host([data-color-scheme="dark"]) {
    --colors-background-primary: oklch(0% 0 0 / 0.5);
}"
`;

exports[`context: [color-scheme] 2`] = `
"export const colorsBackgroundPrimary = 'var(--colors-background-primary)'
"
`;

exports[`context: [media-query, color-scheme] 1`] = `
"@media screen and (min-width: 1920px) {
    :host {
        --size-dynamic-100: 0.2rem;
    }
}

:host([data-media-query="tv"]) {
    --size-dynamic-100: 0.2rem;
}

@media screen and (max-width: 767px) {
    :host {
        --size-dynamic-100: 0.1rem;
    }
}

:host([data-media-query="mobile"]) {
    --size-dynamic-100: 0.1rem;
}

:host {
    --size-dynamic-100: 0.15rem;
    --colors-background-primary: oklch(100% 0 0 / 0.5);
}

@media screen and (min-width: 1024px) {
    :host {
        --size-dynamic-100: 0.15rem;
    }
}

:host([data-media-query="desktop"]) {
    --size-dynamic-100: 0.15rem;
}

@media (prefers-color-scheme:light) {
    :host {
        --colors-background-primary: oklch(100% 0 0 / 0.5);
    }
}

:host([data-color-scheme="light"]) {
    --colors-background-primary: oklch(100% 0 0 / 0.5);
}

@media (prefers-color-scheme:dark) {
    :host {
        --colors-background-primary: oklch(0% 0 0 / 0.5);
    }
}

:host([data-color-scheme="dark"]) {
    --colors-background-primary: oklch(0% 0 0 / 0.5);
}"
`;

exports[`context: [media-query, color-scheme] 2`] = `
"export const sizeDynamic100 = 'var(--size-dynamic-100)'
export const colorsBackgroundPrimary = 'var(--colors-background-primary)'
"
`;

exports[`alias 1`] = `
":host {
    --colors-gray-100: oklch(98.21% 0 0 / none);
}"
`;

exports[`alias 2`] = `
"export const colorsGray100 = 'var(--colors-gray-100)'
export const surfaceSecondary100 = colorsGray100
"
`;

exports[`alias + context 1`] = `
":host {
    --colors-white-100: oklch(97.91% 0 0 / none);
    --colors-charcoal-100: oklch(18.31% 0.004 285.99 / none);
    --surface-secondary-100: var(--colors-white-100);
    --surface-secondary-100: var(--colors-charcoal-100);
}"
`;

exports[`alias + context 2`] = `
"export const colorsWhite100 = 'var(--colors-white-100)'
export const colorsCharcoal100 = 'var(--colors-charcoal-100)'
export const surfaceSecondary100 = 'var(--surface-secondary-100)'
"
`;

exports[`exercise token types 1`] = `
":host {
    --font-weight: 700;
    --letter-spacing: normal;
    --border: 0.05rem solid oklch(0% 0 0 / none);
    --drop-shadow: 4 4 oklch(0% 0 0 / 0.5);
    --gradient: linear-gradient(45deg,oklch(0% 0 0 / 0.5) 20%,oklch(100% 0 0 / 0.5) 80%);
    --font-family: Roboto,sans-serif;
    --transition: 0.3s 0s ease-in-out;
    --transition-cubic: 0.3s 0s cubic-bezier(0.25 0.1 0.25 1);
    --gap: 0.5rem;
    --gap-percent: 10%;
    --grid-template: "a a a" "b c c" "b c c";
    --grid-template-columns: 3rem 3rem;
    --font-size: 0.7rem;
    --line-height: 1;
    --primitive-string: primitive string;
    --primitive-number: 50;
}"
`;

exports[`exercise token types 2`] = `
"export const fontWeight = 'var(--font-weight)'
export const letterSpacing = 'var(--letter-spacing)'
export const border = 'var(--border)'
export const dropShadow = 'var(--drop-shadow)'
export const gradient = 'var(--gradient)'
export const fontFamily = 'var(--font-family)'
export const transition = 'var(--transition)'
export const transitionCubic = 'var(--transition-cubic)'
export const gap = 'var(--gap)'
export const gapPercent = 'var(--gap-percent)'
export const gridTemplate = 'var(--grid-template)'
export const gridTemplateColumns = 'var(--grid-template-columns)'
export const fontSize = 'var(--font-size)'
export const lineHeight = 'var(--line-height)'
export const typography = {
  fontFamily: fontFamily,
  fontSize: fontSize,
  fontWeight: fontWeight,
  letterSpacing: letterSpacing,
  lineHeight: lineHeight,
}
export const flex = {
  display: ,
  columnGap: gap,
}
export const grid = {
  display: ,
  columnGap: gap,
  rowGap: gapPercent,
}
export const primitiveString = 'var(--primitive-string)'
export const primitiveNumber = 'var(--primitive-number)'
"
`;
